flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -600,61 +603,69 @@ private void allocateSlot(
             ResourceCounter pendingSlots) {
-        for (Map.Entry<ResourceProfile, Integer> missingResource : missingResources.entrySet()) {
+        for (Map.Entry<ResourceProfile, Integer> missingResource : missingResources) {
             ResourceProfile profile = missingResource.getKey();
             for (int i = 0; i < missingResource.getValue(); i++) {
-                if (!tryFulfillWithPendingSlots(profile, pendingSlots)) {
-                    boolean couldAllocateWorkerAndReserveSlot =
+                final Tuple2<Boolean, ResourceCounter> matchingResult =
+                        tryFulfillWithPendingSlots(profile, pendingSlots);
+                pendingSlots = matchingResult.f1;
+                if (!matchingResult.f0) {
+                    final Tuple2<Boolean, ResourceCounter> allocationResult =
                             tryAllocateWorkerAndReserveSlot(profile, pendingSlots);
+                    pendingSlots = allocationResult.f1;
+                    boolean couldAllocateWorkerAndReserveSlot = allocationResult.f0;
                     if (!couldAllocateWorkerAndReserveSlot && sendNotEnoughResourceNotifications) {
                         LOG.warn("Could not fulfill resource requirements of job {}.", jobId);
                         resourceActions.notifyNotEnoughResourcesAvailable(
                                 jobId, resourceTracker.getAcquiredResources(jobId));
-                        return;
+                        return pendingSlots;
                     }
                 }
             }
         }
+        return pendingSlots;
     }
 
-    private boolean tryFulfillWithPendingSlots(
+    private Tuple2<Boolean, ResourceCounter> tryFulfillWithPendingSlots(
             ResourceProfile resourceProfile, ResourceCounter pendingSlots) {
-        Set<ResourceProfile> pendingSlotProfiles = pendingSlots.getResourceProfiles();
+        Set<ResourceProfile> pendingSlotProfiles = pendingSlots.getResources();
 
         // short-cut, pretty much only applicable to fine-grained resource management
         if (pendingSlotProfiles.contains(resourceProfile)) {
-            pendingSlots.decrementCount(resourceProfile, 1);
-            return true;
+            pendingSlots = pendingSlots.subtract(resourceProfile, 1);
+            return Tuple2.of(true, pendingSlots);
         }
 
         for (ResourceProfile pendingSlotProfile : pendingSlotProfiles) {
             if (pendingSlotProfile.isMatching(resourceProfile)) {
-                pendingSlots.decrementCount(pendingSlotProfile, 1);
-                return true;
+                pendingSlots = pendingSlots.subtract(pendingSlotProfile, 1);
+                return Tuple2.of(true, pendingSlots);
             }
         }
 
-        return false;
+        return Tuple2.of(false, pendingSlots);
     }
 
-    private boolean tryAllocateWorkerAndReserveSlot(
+    private Tuple2<Boolean, ResourceCounter> tryAllocateWorkerAndReserveSlot(
```

Comment:
Same here. I would suggest to introduce something like `WorkerAllocationResult` or so.